optuna/samplers/_tpe/sampler.py

```diff
@@ -590,3 +566,29 @@ def _get_observation_pairs(
             values[param_name].append(param_value)
 
     return values, scores
+
+
+def _split_observation_pairs(
+    config_vals: Dict[str, List[Optional[float]]],
+    loss_vals: List[Tuple[float, float]],
+    n_below: int,
+) -> Tuple[Dict[str, np.ndarray], Dict[str, np.ndarray]]:
+
+    # `None` items are intentionally converted to `nan` and then filtered out.
+    # For `nan` conversion, the dtype must be float.
+    config_values = {k: np.asarray(v, dtype=float) for k, v in config_vals.items()}
+    loss_values = np.asarray(loss_vals, dtype=[("step", float), ("score", float)])
+
+    index_loss_ascending = np.argsort(loss_values)
+    # `np.sort` is used to keep chronological order.
+    index_below = np.sort(index_loss_ascending[:n_below])
+    index_above = np.sort(index_loss_ascending[n_below:])
+    below = {}
+    above = {}
+    for param_name, param_val in config_values.items():
+        param_val_below = param_val[index_below]
+        param_val_above = param_val[index_above]
+        below[param_name] = param_val_below[~np.isnan(param_val_below)]
+        above[param_name] = param_val_above[~np.isnan(param_val_above)]
+
+    return below, above
```

Comment:
Could you tell me why you extract the `_split_observation_pairs` as a free function more detail? If it is just for the test purpose, we can create an instance and call the method.

Reply:
This is not only for ease of testing, but also to improve the readability of the code. I think the code will be cleaner by improving the symmetry of `get_observation_pairs` and `split_observation_pairs`.

Reply:
That makes sense. Thank you for your explanation.